streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -463,18 +491,46 @@ abstract class ZStream[-R, +E, +O](val process: ZManaged[R, Nothing, ZIO[R, Opti
 
   /**
    * Allows a faster producer to progress independently of a slower consumer by buffering
-   * up to `capacity` elements in a sliding queue.
+   * up to `capacity` chunks in a sliding queue.
    *
    * @note Prefer capacities that are powers of 2 for better performance.
    */
-  final def bufferSliding(capacity: Int): ZStream[R, E, O] =
+  final def bufferChunksSliding(capacity: Int): ZStream[R, E, O] =
     ZStream {
       for {
         queue <- Queue.sliding[(Take[E, O], Promise[Nothing, Unit])](capacity).toManaged(_.shutdown)
         pull  <- bufferSignal(queue)
       } yield pull
     }
 
+  /**
+   * Allows a faster producer to progress independently of a slower consumer by buffering
+   * up to `capacity` elements in a dropping queue.
+   *
+   * @note Prefer capacities that are powers of 2 for better performance.
+   */
+  final def bufferDropping(capacity: Int): ZStream[R, E, O] =
+    ZStream {
+      for {
+        queue <- Queue.dropping[Exit[Option[E], O]](capacity).toManaged(_.shutdown)
+        pull  <- bufferElement(queue)
+      } yield pull
+    }
+
+  /**
+   * Allows a faster producer to progress independently of a slower consumer by buffering
+   * up to `capacity` elements in a sliding queue.
+   *
+   * @note Prefer capacities that are powers of 2 for better performance.
+   */
+  final def bufferSliding(capacity: Int): ZStream[R, E, O] =
+    ZStream {
+      for {
+        queue <- Queue.sliding[Exit[Option[E], O]](capacity).toManaged(_.shutdown)
+        pull  <- bufferElement(queue)
```

Comment:
You'll need to replicate the `bufferSignal` logic for the `bufferSliding` and `bufferDropping` cases; your `bufferElement` implementation is good for the plain backpressuring queue case, but when the queue is sliding or dropping, it is likely to lose error or end signals.

We should also replicate the testcases so we are guarding against these issues.

Reply:
Good point, will update

Reply:
Hmm, isn't this just a problem for bufferDroping though? In case of bufferSliding, the producer will stop as soon as one error is found therefore it won't be lost. Am I missing something? @iravid

Reply:
@iravid I'm about to reopen this pull request, I'm now wondering how the `bufferElement` implementation won't work for `bufferSliding` and `bufferDropping` if the producer stops as soon as there's an error encountered. How the error or end signal will be lost?

Reply:
For bufferDropping, consider the case where the consumer is slow and the queue is full. When the producer tries to offer an end/error signal, it gets dropped. The producer then stops. The consumer eventually drains the queue and hangs because the queue is empty and nothing is producing to it anymore.

bufferSliding is a different case. In that case, if the consumer is slow, we will be dropping an element that is present in the queue in favour of an error/end signal. This is arguable but I think that's undesirable.

Either way, I think just slapping a BufferedPull on the existing implementation of `bufferSignal` could do the trick with a minimal amount of work.

Reply:
And thank you for following up!